functions/src/sections/Section.service.ts

```diff
@@ -26,24 +26,57 @@ export class SectionsService {
     code: string
   ): Promise<Seat[]> {
     // get term, subject, code to crn mappings from db.
-    const doc = await db.sectionMappings
+    const mapping =
+      (await this.getSectionMapping(term, subject, code)) ||
+      (await this.setSectionMapping(term, subject, code));
+
+    if (mapping) {
+      return await Promise.all(
+        mapping.crns.map(async (crn) => {
+          const seat = await UVicCourseScraper.getSectionSeats(term, crn);
+          return { ...seat, crn, date: new Date(Date.now()) };
+        })
+      );
+    }
+    throw new SectionNotFoundError('Seats Not Found');
+  }
+
+  public async getSectionMapping(
+    term: string,
+    subject: string,
+    code: string
+  ): Promise<CourseMapping | undefined> {
+    const doc = await db.courseMappings
       .doc(SectionsService.constructSectionKey(term, subject, code))
       .get();
 
-    const data = doc.data();
-
-    if (!data) {
-      throw new SectionNotFoundError('Seats Not Found');
+    const t = doc.data()?.retrieveAt?.getTime();
+    // if retrivedAt exists and it was retieved within 30 minutes
+    if (t && t + 1000 * 1800 > Date.now()) {
+      return undefined;
     }
+    return doc.data();
+  }
 
-    const seats = await Promise.all(
-      data.crns.map(async (crn) => {
-        const seat = await UVicCourseScraper.getSectionSeats(term, crn);
-        return { ...seat, crn };
-      })
-    );
-
-    return seats;
+  public async setSectionMapping(
+    term: string,
+    subject: string,
+    code: string
+  ): Promise<CourseMapping | void> {
+    try {
+      const sections = await UVicCourseScraper.getCourseSections(
+        term,
+        subject,
+        code
+      );
+      if (sections.length > 0) {
+        const crns = sections.map(({ crn }) => crn);
+        await db.courseMappings
+          .doc(SectionsService.constructSectionKey(term, subject, code))
+          .set({ crns });
+        return { crns, retrieveAt: new Date(Date.now()) };
+      }
+    } catch (e) {}
   }
 
   public async updateSectionMappings(): Promise<void> {
```

Comment:
should we just delete this then?

Reply:
I don't think so. I can see this being useful in some situations or tasks during development.

Reply:
we could just delete it here and then copy it in from the git history if we ever need it so we don't have dead code...

Reply:
While we're still experimenting with caching so I'd prefer to leave it in.